file/builder.go

```diff
@@ -768,9 +786,101 @@ func (b *stateBuilder) ingestRoute(r FRoute) error {
 		p.Route = &kong.Route{ID: kong.String(*r.ID)}
 		plugins = append(plugins, *p)
 	}
+	plugins, err = b.ingestPluginDefaults(plugins)
+	if err != nil {
+		return err
+	}
 	return b.ingestPlugins(plugins)
 }
 
+func ingestDefaultConfigFields(p *FPlugin, config interface{}) {
+	for _, field := range config.([]interface{}) {
+		for key, value := range field.(map[string]interface{}) {
+			if p.Config[key] != nil {
+				continue
+			}
+			if defaultValue, ok := value.(map[string]interface{})["default"]; ok {
+				p.Config[key] = defaultValue
+			} else if fields, ok := value.(map[string]interface{})["fields"]; ok {
+				innerField := map[string]interface{}{}
+				for _, field := range fields.([]interface{}) {
+					for fieldKey, value := range field.(map[string]interface{}) {
+						if defaultValue, ok := value.(map[string]interface{})["default"]; ok {
+							innerField[fieldKey] = defaultValue
+						} else {
+							innerField[fieldKey] = nil
+						}
+					}
+				}
+				p.Config[key] = innerField
+			} else {
+				p.Config[key] = nil
+			}
+		}
+	}
+}
+
+func ingestDefaultProtocols(p *FPlugin, protocols interface{}) {
+	if defaultValue, ok := protocols.(map[string]interface{})["default"]; ok {
+		for _, protocol := range defaultValue.([]interface{}) {
+			if protocolStr, ok := protocol.(string); ok {
+				p.Protocols = append(p.Protocols, &protocolStr)
+			}
+		}
+	}
+}
+
+func (b *stateBuilder) getDefaults(entity, entityID string) (map[string]interface{}, error) {
+	ctx := context.Background()
+	var schema map[string]interface{}
+	endpoint := fmt.Sprintf("/schemas/%s", entity)
+	if entityID != "" {
+		endpoint += fmt.Sprintf("/%s", entityID)
+	}
+	req, err := b.client.NewRequest(http.MethodGet, endpoint, nil, nil)
+	if err != nil {
+		return schema, err
+	}
+	_, err = b.client.Do(ctx, req, &schema)
+	return schema, err
+}
+
+func (b *stateBuilder) ingestPluginDefaults(plugins []FPlugin) ([]FPlugin, error) {
+	// skip for konnect and when kong client is not defined (like with validate)
+	if b.konnectRawState == nil ||
+		len(b.konnectRawState.Documents) > 0 ||
+		len(b.konnectRawState.ServicePackages) > 0 ||
+		b.client == nil {
+		return plugins, nil
+	}
+	pluginsWithDefault := []FPlugin{}
+	for _, p := range plugins {
+		p := p
+		defaults, err := b.getDefaults("plugins", *p.Name)
```

Comment:
Please introduce caching of plugin defaults.
If there are 100 key-auth plugins defined by the user, this code will invoke a GET request for the same endpoint (and same response) 100 times. This will have a severe impact on execution time of decK.